test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadState/GetThreadStateTest.java

```diff
@@ -0,0 +1,422 @@
+ */
+
+import java.lang.ref.Reference;
+import java.util.StringJoiner;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.LockSupport;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+class GetThreadStateTest {
+
+    @BeforeAll
+    static void setup() {
+        System.loadLibrary("GetThreadStateTest");
+        init();
+    }
+
+    /**
+     * Test state of new/unstarted thread.
+     */
+    @Test
+    void testUnstarted() {
+        var thread = Thread.ofVirtual().unstarted(() -> { });
+        check(thread, /*new*/ 0);
+    }
+
+    /**
+     * Test state of terminated thread.
+     */
+    @Test
+    void testTerminated() throws Exception {
+        var thread = Thread.ofVirtual().start(() -> { });
+        thread.join();
+        check(thread, JVMTI_THREAD_STATE_TERMINATED);
+    }
+
+    /**
+     * Test state of runnable thread.
+     */
+    @Test
+    void testRunnable() throws Exception {
+        var latch = new CountDownLatch(1);
+        var done = new AtomicBoolean();
+        var thread = Thread.ofVirtual().start(() -> {
+            latch.countDown();
+
+            // spin until done
+            while (!done.get()) {
+                Thread.onSpinWait();
+            }
+        });
+        try {
+            // wait for thread to start execution
+            latch.await();
+
+            // thread should be runnable
+            int expected = JVMTI_THREAD_STATE_ALIVE | JVMTI_THREAD_STATE_RUNNABLE;
+            check(thread, expected);
+
+            // re-test with interrupt status set
+            thread.interrupt();
+            check(thread, expected | JVMTI_THREAD_STATE_INTERRUPTED);
+        } finally {
+            done.set(true);
+            thread.join();
+        }
+    }
+
+    /**
+     * Test state of thread waiting to enter a monitor.
+     */
+    @Test
+    void testMonitorEnter() throws Exception {
+        var latch = new CountDownLatch(1);
+        Object lock = new Object();
+        var thread = Thread.ofVirtual().unstarted(() -> {
+            latch.countDown();
+            synchronized (lock) { }
+        });
+        try {
+            synchronized (lock) {
+                // start thread and wait for it to start execution
+                thread.start();
+                latch.await();
+
+                // thread should block on monitor enter
+                int expected = JVMTI_THREAD_STATE_ALIVE | JVMTI_THREAD_STATE_BLOCKED_ON_MONITOR_ENTER;
+                await(thread, expected);
+
+                // re-test with interrupt status set
+                thread.interrupt();
+                check(thread, expected | JVMTI_THREAD_STATE_INTERRUPTED);
+            }
+        } finally {
+            thread.join();
+            Reference.reachabilityFence(lock);
```

Comment:
It would be nice to add a short comment why this is needed.

Reply:
> It would be nice to add a short comment why this is needed.

It's not needed here, it was needed in another version of the tests. While benign, I can remove it to avoid any confusion. Thanks.

Reply:
There are more such lines: 176, 219, 313,